fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/principal/JWTCallerPrincipalFactory.java

```diff
@@ -84,17 +84,19 @@ private static JWTCallerPrincipalFactory loadSpi(ClassLoader cl) {
             URL u = cl.getResource("/META-INF/services/org.eclipse.microprofile.jwt.principal.JWTCallerPrincipalFactory");
             log.debugf("loadSpi, cl=%s, u=%s, sl=%s", cl, u, sl);
             try {
-                for (JWTCallerPrincipalFactory spi : sl) {
-                    if (instance != null) {
-                        throw new IllegalStateException(
-                                "Multiple JWTCallerPrincipalFactory implementations found: "
-                                        + spi.getClass().getName() + " and "
-                                        + instance.getClass().getName());
-                    } else {
-                        log.debugf("sl=%s, loaded=%s", sl, spi);
-                        instance = spi;
+                for (Object spi : sl) {
+                    if (spi instanceof JWTCallerPrincipalFactory) {
+                        if (instance != null) {
+                            throw new MultipleFactoriesException(spi.getClass().getName());
+                        } else {
+                            log.debugf("sl=%s, loaded=%s", sl, spi);
+                            instance = (JWTCallerPrincipalFactory)spi;
+                        }
                     }
                 }
+            } catch (MultipleFactoriesException e) {
+                log.warn("Multiple JWTCallerPrincipalFactory implementations found: "
```

Comment:
What do we benefit from throwing the exception and catching it instead of printing this warning directly?

Reply:
Sorry, missed this comment. I introduced it initially when trying to avoid the ambiguity at the catch level so that not to log harmless warnings which is less relevant now with the PR update (after Martin pointed out at the lazy ServiceLoader.iterator behavior).
During the 2nd update I was actually thinking about reverting the new exception class as well, but  decided that with the dedicated exception there will be a single warning only in case of the multiple factories issue as opposed to two if I revert it (within the for loop and then in the catch - where it would also be somewhat confusing because the 2nd warning would say a provider failed to be loaded which would not be true in case of the multiple factories issue). 
Does it sound reasonable ?

Reply:
The original code does not warn within the loop, it only throws ISE which is then logged. I can revert to that, though again a user would see something like "Failed to load the provider: multiple implementations found" which is somewhat contradictory. Either way, I can either revert the new class or keep it :-)

Reply:
It can stay as it is, it just looks to me that the code below would have the same effect with a bit less code.
But maybe I'm missing sth :)
```
if (instance != null) {
   log.warn("Multiple JWTCallerPrincipalFactory implementations found: " + spi.getClass().getName() and ...")
   break;
}
```

Reply:
Sure, sounds good to me, 'break' works well. If Martin does not object then I'll remove the newly introduced exception and simplify. Thanks :-)

Reply:
+1